CMakeLists.txt

```diff
@@ -1,17 +1,84 @@
 cmake_minimum_required(VERSION 3.10)
 
-project(cxxeffect)
+project(cxxeffect CXX)
 
-# add the executable
-add_executable(cxxeffect
-    src/stream.hpp
-    src/stream.cpp
-    src/task.hpp
-    src/task.cpp
-    src/eff.hpp
-    src/eff.cpp)
+# We want to export compile commands b/c it's useful for ccls
+set(CMAKE_EXPORT_COMPILE_COMMANDS true)
 
-set_property(TARGET cxxeffect PROPERTY CXX_STANDARD 17)
+# cxxeffect is an interface library b/c it has a lot of templated code
+# That means it's best implemented as header-only without the need for cpp files
+add_library(cxxeffect INTERFACE)
 
-target_link_libraries(cxxeffect pthread)
+set_property(TARGET cxxeffect PROPERTY CXX_STANDARD_REQUIRED 20)
 
+target_compile_features(cxxeffect INTERFACE cxx_std_20)
+
+# boilerplate compiler options to have support for C++20. We can remove if you don't want them.
+if (MSVC)
+    target_compile_options(cxxeffect INTERFACE "/EHsc")
+endif()
+
+
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" )
+    target_compile_options(cxxeffect INTERFACE "-fcoroutines")
+endif()
+
+
+target_compile_definitions(cxxeffect INTERFACE)
+
+target_include_directories(
+    cxxeffect
+    INTERFACE
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include/>
+    $<INSTALL_INTERFACE:include>
+)
+
+# Provides install target, so users can install the library on their system
+# if they like
+
+option(CXXEFFECT_Install "Install CMake targets during install step." ON)
+install(
+    DIRECTORY ${PROJECT_SOURCE_DIR}/include/
+    DESTINATION include
+)
+
+# Use pthreads, if availible
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+# Enable threads in project
+find_package(Threads REQUIRED)
+
+# If we're building this project, and not something that has this project as a dependency
+# Then we want to build files in the examples directory (which contains the cat)
+# This will compile every file in the examples directory as a separate program
+if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
```

Comment:
What do you think about a flag `CXXEFFECT_BUILD_EXAMPLES` that's by-default `ON`, and gets &&ed with this?

Reply:
I've updated all_tests to all_examples, and added documentation for the "/EH" flag. I think some of these changes and improvements should be made on a separate branch specifically for the build system. Do you think that would work?